Extracted PlatformViewsChannel from PlatformViewsController.

matthew-carroll opened the pull request:
This PR extracts all platform view system channel communication into `PlatformViewsChannel`.

The relationship between `PlatformViewsController` and `PlatformViewsChannel` is analogous to multiple other relationships including: `TextInputPlugin` and `TextInputChannel`, `AndroidKeyProcessor` and `KeyEventChannel`, and `AccessibilityBridge` and `AccessibilityChannel`.

The idea behind these relationships is that the channels codify the Android/Flutter communication API, while the other associated class handles any state and handler behavior.

At a technical level, this PR handles communication issues with `Exception`s so that the inner workings of method channels are completely hidden from down stream consumers.

amirh commented on shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformViewsChannel.java:
```diff
@@ -0,0 +1,479 @@
+
+/**
+ * System channel that sends 2-way communication between Flutter and Android to
+ * display Android Views within Flutter.
```
This channel and the PlatformViewsController isn't displaying the views, it configures them, and passes input events.

Maybe replace "to display Android Views within Flutter" with "to facilitate embedding of Android Views within a Flutter application".

matthew-carroll replied:
done

amirh commented on shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformViewsChannel.java:
```diff
@@ -0,0 +1,479 @@
+   * An attempt was made to use platform views on a version of Android that platform
+   * views does not support.
+   */
+  public static class InvalidPlatformVersionException extends IllegalStateException {}
```
I'd favor just throwing IllegalStateException directly with an error message, these specialized exception types are feeling a little too boiler plate-y to me. There's a more detailed explanation on why it's good to reuse IllegallStateException directly in Effective Java item 72. Sending you the internal best practices link that explains it offline.

matthew-carroll replied:
done

amirh commented on shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:
```diff
@@ -47,11 +43,51 @@
 
-    // The messenger used to communicate with the framework over the platform views channel.
-    private BinaryMessenger mMessenger;
+    // The system channel used to communicate with Flutter about platform views.
```
I'd keep it saying: to communicate with "the framework", this code is also part of Flutter.

matthew-carroll replied:
done

amirh commented on shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:
```diff
@@ -47,11 +43,51 @@
 
     private final HashMap<Integer, VirtualDisplayController> vdControllers;
 
+    private final PlatformViewsChannel.PlatformViewsHandler channelHandler = new PlatformViewsChannel.PlatformViewsHandler() {
```
What's the benefit of having this anonymous class over making PlatformViewsController implement the interface?

matthew-carroll replied:
The distinction is based on the intended use. Implementing an interface is a public statement that the given instance can be used as the interface. In other words, a PlatformViewsController could be used by client code as a PlatformViewsHandler.  But if another area of code used this class as a PlatformViewsHandler, that would very likely break this code. What we really want is for this PlatformViewsController to own both a PlatformViewsChannel, as well as a PlatformViewsHandler. These are internal implementation details of a PlatformViewsController, rather than a public behavior of PlatformViewsController. Therefore, the PlatformViewsHandler is kept private and internal.

amirh replied:
I see.

What do you think about keeping the actual method implementations in this anonymous class (e.g inlining `PlatformViewsController#createPlatformView` here).

We're adding quite a lot of indirection layers here, I'm convinced that all of them are justified but this one(see my note to self in the PR comment).

matthew-carroll replied:
So the recommendation is to avoid re-declaring each method within the outer class and to simply do the behavior within the anonymous class?

If so, then I think that should be fine. I think the only reason I didn't do that was because I was trying to create a minimal diff against the existing code. I wasn't sure at the time if you were in the middle of evolving this class. 

So if you confirm that this change is a good resolution then I'll take care of it.

amirh replied:
Yeah I support inlining the implementation if there isn't a good reason to keep the methods separate.

amirh commented on shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformViewsChannel.java:
```diff
@@ -0,0 +1,479 @@
+        return;
+      }
+
+      switch (call.method) {
```
This method is pretty long which makes it hard for me to follow, lets keep split out the handling of the each different switch case to its own separate methods.

matthew-carroll replied:
done

cbracken commented:
@matthew-carroll did you have plans to address comments from @amirh or should we close this PR?

matthew-carroll commented:
@cbracken I plan to address them

cbracken commented:
@matthew-carroll any progress on this?

matthew-carroll commented:
I'm gonna try to refresh this PR this weekend and hopefully have it reviewable by Monday.

matthew-carroll commented:
@amirh I believe I've addressed your comments. I had to force push an update so that I could get the latest from master, and also because the first 2 commits had the wrong author info for me.

amirh commented:
Coming back to this PR 3 months later it's almost like taking a fresh look:

I like the separation of parsing the method channel message from handling it, I think it makes the code easier to follow when the parsing is complex.

I was a little about the amount of extra indirection layers and the boilerplate that comes with them, but have convinced myself it makes sense, so a note to future-self as for why we're adding all this boilerplate:

If I want to add a new `foo` method to the method channel with this change I need to:
1.  add a switch case in the `MethodCallHandler` implementation.
2. implement `PlatformViewsChannel#foo`.
3. add `foo` to the `PlatformViewsHandler`.
4. add a `foo` implementation in the anonymous implementation of `channelHandler`.
5. add a private `foo` implementation in PlatformViewsController.

The separation between 1<-->2 we want to avoid a huge `onMethodCall`method.
`#3` we want to avoid a cyclic dependency between `PlatformViewsController` and `PlatformViewsHandler`.
The separation between 4<-->5 I'm not yet convinced we want, I'm asking to merge them in this review iteration.

matthew-carroll commented:
@amirh I rebased to fix some conflicts and then I inlined the channel handler implementation as you suggested. Are we good to merge in at this point?
